pkg/placement/hash-domain.go

```diff
@@ -0,0 +1,170 @@
+/*
+Copyright 2023 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package placement
+
+import (
+	"hash/crc64"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	edgeapi "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
+)
+
+type HashDomain[Elt any] interface {
+	Equal(Elt, Elt) bool
+	Hash(Elt) HashValue
+}
+
+type HashValue = uint64
+
+func NewHashDomainFuncs[Elt any](equal func(Elt, Elt) bool, hash func(Elt) HashValue) HashDomainFuncs[Elt] {
+	return HashDomainFuncs[Elt]{equal, hash}
+}
+
+type HashDomainFuncs[Elt any] struct {
+	DoEqual func(Elt, Elt) bool
+	DoHash  func(Elt) HashValue
+}
+
+func (hdf HashDomainFuncs[Elt]) Equal(left, right Elt) bool { return hdf.DoEqual(left, right) }
+func (hdf HashDomainFuncs[Elt]) Hash(elt Elt) HashValue     { return hdf.DoHash(elt) }
+
+func PairHashDomain[First, Second any](first HashDomain[First], second HashDomain[Second]) HashDomain[Pair[First, Second]] {
+	return HashDomainFuncs[Pair[First, Second]]{
+		DoEqual: func(left, right Pair[First, Second]) bool {
+			return first.Equal(left.First, right.First) && second.Equal(left.Second, right.Second)
+		},
+		DoHash: func(tup Pair[First, Second]) HashValue { return first.Hash(tup.First)*257 + second.Hash(tup.Second) },
+	}
+}
+
+func TripleHashDomain[First, Second, Third any](first HashDomain[First], second HashDomain[Second], third HashDomain[Third]) HashDomain[Triple[First, Second, Third]] {
+	return HashDomainFuncs[Triple[First, Second, Third]]{
+		DoEqual: func(left, right Triple[First, Second, Third]) bool {
+			return first.Equal(left.First, right.First) && second.Equal(left.Second, right.Second) && third.Equal(left.Third, right.Third)
+		},
+		DoHash: func(tup Triple[First, Second, Third]) HashValue {
+			return first.Hash(tup.First)*65539 + second.Hash(tup.Second)*257 + third.Hash(tup.Third)
```

Comment:
Not sure my understanding is correct or not. Given `const numHashBuckets = 8`, is this line effectively equivalent to `return first.Hash(tup.First)*3 + second.Hash(tup.Second)*1 + third.Hash(tup.Third)`?

Reply:
Yes.  I do not want to spend time making great hash functions right now, that is not the priority.  I suspect there must be a decent hash map package that we could import, but I also do not want to spend time looking for it now.  We just need something that works well enough for testing this week.

Reply:
I see. Thanks!